tests/functional/base.py

```diff
@@ -0,0 +1,316 @@
+import os
+import configparser
+import tempfile
+import json
+from tempest.lib.cli import base
+from tempest.lib.cli import output_parser
+from tempest.lib.common.utils import data_utils
+
+class EsiFunctionalBase(base.ClientTestBase):
+    @classmethod
+    def setUpClass(cls):
+        super(EsiFunctionalBase, cls).setUpClass()
+        cls.clients = {}
+        cls.config = {}
+        cls.client_info = {}
+        cls._cleanups = []
+
+        cls._init_config(cls)
+        cls._init_client(cls, 'admin')
+        cls._init_roles(cls)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls._cleanups.reverse()
+        for cleanup in cls._cleanups:
+            cleanup[0](*cleanup[1:])
+
+    def _get_clients(self):
+        #NOTE: ClientTestBase's constructor requires this to be implemented, but
+        #to initialize our clients, we need ClientTestBase's constructor to have
+        #been called. We return nothing and leave the job up to _init_clients().
+        return {}
+
+    def _init_config(self):
+        venv_name = os.environ.get('OS_VENV_NAME', default='functional')
+        self.config['cli_dir'] = os.path.join(os.path.abspath('.'), 
+                '.tox/%s/bin' % venv_name)
+
+        # allow custom configuration to be passed in via env var
+        cfg_file_path = os.environ.get('OS_ESI_CFG_PATH',
+                '/etc/esi-leap/esi-leap.conf')
+        cfg_parser = configparser.ConfigParser()
+        if not cfg_parser.read(cfg_file_path):
+            self.fail('Could not open config file %s for reading' %
+                    cfg_file_path)
+            return -1
+
+        # attempts to read authentication credentials in this order:
+        # 1) [keystone] section of config file
+        # 2) [keystone_authtoken] section of config file
+        # 3) from the environment variables (e.g. OS_PASSWORD, etc)
+        def auth_get(val):
+            for sect in 'keystone', 'keystone_authtoken':
+                if cfg_parser.has_section(sect):
+                    try:
+                        return cfg_parser.get(sect, val)
+                    except:
+                        pass
+            x = os.environ.get('OS_%s' % val.upper())
+            if x != None:
+                return x
+            else:
+                raise configparser.NoOptionError
+
+        admin_auth = {}
+        try:
+            if cfg_parser.has_section('dummy_node'):
+                opt = 'dummy_node_dir'
+                self.config[opt] = cfg_parser.get('dummy_node', opt)
+            # TODO: add support for keystone v2 (this is v3-only atm)
+            auth_opts = ['auth_type', 'auth_url', 'username', 'password',
+                    'project_name', 'user_domain_name', 'project_domain_name']
+            for opt in auth_opts:
+                admin_auth[opt] = auth_get(opt)
+        except (configparser.NoOptionError):
+            self.fail("Missing option %s in configuration file." % opt)
+            return -1
```

Comment:
If `self.fail` raises an exception, then the code won't reach this return statement anyway. you probably don't need to return 0 or -1 in most of the cases here unless those numbers mean something to the caller.

Reply:
These return statements are indeed meaningless (this was the first chunk of code I wrote for the base class and my mind was still stuck in C)